crates/ruff_linter/src/rules/pyupgrade/rules/printf_string_formatting.rs

```diff
@@ -286,6 +291,55 @@ fn clean_params_dictionary(right: &Expr, locator: &Locator, stylist: &Stylist) -
     Some(contents)
 }
 
+/// (In the case where the format string has only a single %-placeholder) and the righthand operand
+/// to % is a variable, try to offer a fix.
+///
+/// If the variable is not a tuple, then offer `({})`.
+/// ```python
+/// x = set(["hi", "hello"])
+/// print("%s" % x)
+/// print("{}".format(x))
+/// ```
+///
+/// If the variable is a 1-tuple, then offer `({}[0])`.
+/// ```python
+/// x = (1,)
+/// print("%s" % x)
+/// print("{}".format(x[0]))
+/// ```
+///
+/// If the variable is an n-tuple, the user has a bug in their code. Offer no fix.
+fn clean_params_variable<'a>(checker: &mut Checker, right: &Expr) -> Option<Cow<'a, str>> {
+    let Expr::Name(ast::ExprName { id, .. }) = right else {
+        return None;
+    };
+    let semantic = checker.semantic();
+    let binding_id = semantic.lookup_symbol(id.as_str())?;
+    let binding = semantic.binding(binding_id);
+    let rt = checker.locator().slice(right);
+    if let Some(expr) = analyze::typing::find_binding_value(binding, semantic) {
+        match expr {
+            // n-tuple
+            Expr::Tuple(ast::ExprTuple { elts, .. }) if 1 < elts.len() => {
+                return None;
+            }
+            // 1-tuple
+            Expr::Tuple(ast::ExprTuple { elts, .. }) if 1 == elts.len() => {
+                return Some(Cow::Owned(format!("({rt}[0])")));
+            }
+            // non-tuple
+            _ => {
+                return Some(Cow::Owned(format!("({rt})")));
+            }
+        }
+    }
+    // when find_binding_value doesn't work, fall back to offering a fix only for non-tuples
+    if analyze::typing::is_tuple(binding, semantic) {
```

Comment:
I think this will only return `true` if the value is definitively a tuple, but if the value can't be determined, it _could_ still be a tuple. Imagine, e .g., a function argument with no annotation.

Reply:
In other words, couldn't the value still be a tuple, even if `is_tuple` returns `false`? I'm actually not sure why we aren't seeing more fixes in the ecosystem checks given this.

Reply:
Yeah, it seems odd. The code using `find_binding_value` should not be suceeding in all cases, so we should be seeing the behavior fall back to `is_tuple` sometimes, and therefore offer that fix (wrongly) in some of those cases. :thinking:

Reply:
Yeah, the behavior is matching my expectation.

```python
zzz = "abc"
'Hello %s' % zzz
```

Here, `find_binding_value` returns a string, so we hit line 332 and still offer a fix.

For cases in which `find_binding_value` returns `None`, like:

```python
zzz = foo()
'Hello %s' % zzz
```

We're also offering a fix, because `analyze::typing::is_tuple` returns `false` there (we don't know if `foo()` is or isn't a tuple).

Reply:
It seems that dereffing the variable and/or trying to analyze its type is not worth the additional code.

Reply:
Code won't exist in the next version that I push.